hivemind/dht/__init__.py

```diff
@@ -228,5 +230,8 @@ async def _first_k_active(
                     uid_prefixes[pre_dispatch_chunk_i * chunk_size: (pre_dispatch_chunk_i + 1) * chunk_size],
                     num_workers=num_workers_per_chunk)))
 
+        for task in pending_tasks:
+            task.cancel()
+
         # return k active prefixes or as many as we could find
-        future.set_result(active_prefixes)
+        future.set_result(OrderedDict(found))
```

Comment:
Technically speaking, all dicts are ordered since python 3.7, but i want to emphasize it here for the reader.

Reply:
1. The requirement of returning objects in order defined by arguments holds in many other functions (e.g. get_experts, find_nearest_nodes), but we don't hint at the order there simply because it is clear from the signature
2. Even if this case is different (not all prefixes may be returned), the way this restriction is imposed now is too implicit: the last thing the reader is going to think is "aha, we create OrderedDict, because the order of keys has to be consistent with `uid_prefixes` and not because OrderedDict has methods that `dict` doesn't / we will need order-sensitive comparison / other reasons"
3. Lastly, even with OrderedDict being created for the result, it is easy to break the assumption without anybody noticing

Thus, I suggest one of the following adjustments:
1. We can state the order preserving property in the docstring of first_k_active
2. We can avoid this assumption and restore the order in moe/beam_search (if it is really required). FYI I added `random.shuffle(beam)` after `beam = list(best_alive_prefixes.keys())` in moe.py and the tests still pass